docs/architecture/adr-008-event-subscription.md

```diff
@@ -0,0 +1,178 @@
+subscription][tm-event-subs] RPC endpoint to allow clients to receive
+notifications of specific events as they happen (currently via a WebSockets
+connection). We need to expose this subscription mechanism to the RPC client.
+
+In order to achieve this, we need:
+
+1. An ergonomic interface for the RPC client that allows developers to subscribe
+   to events produced by specific queries. Specifically, this interface must:
+   1. Offer **subscription** functionality, where:
+      1. A single **subscription** takes place to the **events** produced by a
+         **query** (the [PEG] for which, at the time of this writing, is located
+         [here][query-peg]).
+      2. Callers must be able to create multiple distinct subscriptions.
+      3. A subscription must offer an interface to allow for iteration over
+         incoming events only relevant to that subscription (i.e. it should
+         produce an **event stream**).
+      4. It must be possible to, from outside of the RPC client, receive events
+         from multiple subscriptions' event streams concurrently without
+         interfering with/blocking each other.
+      5. It must be possible to, from outside of the RPC client, handle
+         subscriptions' event streams without blocking other RPC operations.
+   2. Offer the ability to **unsubscribe** from specific queries (i.e. to
+      terminate specific subscriptions).
+2. A pub/sub architecture (with an appropriate concurrency model) that
+   facilitates the subscription interface.
+
+## Decision
+
+### Assumptions
+
+* All blocking operations that deal with I/O must be `async`
+* We will not be ["de-asyncifying" the RPC][issue-318] and will rather, in a
+  future ADR, propose a synchronous architecture as well should we need one.
+
+### Proposed Entities and Relationships
+
+The entities in the diagram below are described in the following subsections.
+The diagram also gives some indication as to the proposed concurrency model for
+the architecture and how the entities relate to each other.
+
+![](assets/rpc-client-subscription-rels.png)
+
+### Subscription
+
+A `Subscription` here is envisaged as an entity that implements the
+[Stream][futures-stream] trait, allowing its owner to asynchronously iterate
+through all of the **relevant** incoming events. Use of such a subscription
+should be as simple as:
+
+```rust
+while let Some(result_event) = subscription.next().await {
+   match result_event {
+      Ok(event) => { /* handle event */ },
+      Err(e) => { /* terminate subscription and report error */ },
+   }
+}
+```
+
+Since the underlying events may not be produced at the same cadence as that of
+the `Subscription` owner's processing, it is proposed here to make use of an
+appropriate type of buffered channel to realize this. The size of this buffer
+must be configurable through the RPC Client.
+
+### RPC Pub/Sub
+
+The RPC Client must asynchronously spawn a component `PubSub` whose
```

Comment:
It ought to be the responsibility of the user/caller to spawn this "run loop". Otherwise we take away control about lifetime. Instead of having the client silently spawn things in the background it should be made explicit through its interface.

Reply:
I've implemented a handle/driver architecture for most of the parts of the system that need to operate in two async contexts. The "handle" part allows for total control of the "driver" part.

Unfortunately, carelessly dropping the handle part right now with the current architecture still leaves the driver dangling (which will hopefully be more easily addressed when [async destructors](https://internals.rust-lang.org/t/asynchronous-destructors/11127) are available in Rust).